src/dbnode/client/session.go

```diff
@@ -3476,61 +3526,117 @@ const enqueueChannelDefaultLen = 32768
 func newEnqueueChannel(m *streamFromPeersMetrics) enqueueChannel {
 	c := &enqueueCh{
 		peersMetadataCh: make(chan []receivedBlockMetadata, enqueueChannelDefaultLen),
-		closed:          0,
 		metrics:         m,
 	}
+
 	// Allocate the enqueue delayed fn just once
-	c.enqueueDelayedFn = func(peersMetadata []receivedBlockMetadata) {
+	c.enqueueDelayedFn = func(peersMetadata []receivedBlockMetadata) error {
+		c.Lock()
+		if c.closed {
+			c.Unlock()
+			return errEnqueueChIsClosed
+		}
+		c.sending++
+		c.Unlock()
 		c.peersMetadataCh <- peersMetadata
+		c.Lock()
+		c.sending--
+		c.Unlock()
+		return nil
 	}
+
 	go func() {
-		for atomic.LoadInt64(&c.closed) == 0 {
-			m.blocksEnqueueChannel.Update(float64(len(c.peersMetadataCh)))
+		for {
+			c.Lock()
+			closed := c.closed
+			numEnqueued := float64(len(c.peersMetadataCh))
+			c.Unlock()
+			if closed {
+				return
+			}
+			m.blocksEnqueueChannel.Update(numEnqueued)
 			time.Sleep(gaugeReportInterval)
 		}
 	}()
 	return c
 }
 
-func (c *enqueueCh) enqueue(peersMetadata []receivedBlockMetadata) {
-	atomic.AddUint64(&c.enqueued, 1)
+func (c *enqueueCh) enqueue(peersMetadata []receivedBlockMetadata) error {
+	c.Lock()
+	if c.closed {
+		c.Unlock()
+		return errEnqueueChIsClosed
+	}
+	c.enqueued++
+	c.sending++
+	c.Unlock()
 	c.peersMetadataCh <- peersMetadata
+	c.Lock()
+	c.sending--
+	c.Unlock()
+	return nil
 }
 
-func (c *enqueueCh) enqueueDelayed(numToEnqueue int) func([]receivedBlockMetadata) {
-	atomic.AddUint64(&c.enqueued, uint64(numToEnqueue))
-	return c.enqueueDelayedFn
+func (c *enqueueCh) enqueueDelayed(numToEnqueue int) (func([]receivedBlockMetadata) error, error) {
+	c.Lock()
+	if c.closed {
+		c.Unlock()
+		return nil, errEnqueueChIsClosed
+	}
+	c.enqueued += (numToEnqueue)
+	c.Unlock()
+	return c.enqueueDelayedFn, nil
 }
 
-func (c *enqueueCh) get() <-chan []receivedBlockMetadata {
-	return c.peersMetadataCh
+func (c *enqueueCh) get() (<-chan []receivedBlockMetadata, error) {
+	c.Lock()
+	if c.closed {
+		c.Unlock()
+		return nil, errEnqueueChIsClosed
+	}
+	c.Unlock()
+	metadataCh := c.peersMetadataCh
+	return metadataCh, nil
 }
 
 func (c *enqueueCh) trackPending(amount int) {
-	atomic.AddUint64(&c.enqueued, uint64(amount))
+	c.Lock()
+	c.enqueued += amount
+	c.Unlock()
 }
 
 func (c *enqueueCh) trackProcessed(amount int) {
-	atomic.AddUint64(&c.processed, uint64(amount))
+	c.Lock()
+	c.processed += amount
+	c.Unlock()
 }
 
 func (c *enqueueCh) unprocessedLen() int {
-	return int(atomic.LoadUint64(&c.enqueued) - atomic.LoadUint64(&c.processed))
+	c.Lock()
+	unprocessed := c.unprocessedLenWithLock()
+	c.Unlock()
+	return unprocessed
+}
+
+func (c *enqueueCh) unprocessedLenWithLock() int {
+	return c.enqueued - c.processed
 }
 
 func (c *enqueueCh) closeOnAllProcessed() {
 	defer func() {
-		atomic.StoreInt64(&c.closed, 1)
+		c.Lock()
+		c.closed = true
```

Comment:
Hm, shouldn't this be set while holding the lock `close(c.peersMetadataCh)` is closed?

Otherwise there's a race where you close it but the `c.closed=true` doesn't get updated till a moment after that.